src/group_theory/finiteness.lean

```diff
@@ -0,0 +1,77 @@
+/-
+Copyright (c) 2021 Riccardo Brasca. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Riccardo Brasca
+-/
+
+import data.set.finite
+import group_theory.submonoid.operations
+
+/-!
+# Finitely generated monoid.
+
+We define finitely generated monoids.
+
+## Main definition
+
+* `monoid.fg`, `add_monoid.fg` : A monoid is finitely generated.
+
+-/
+
+section monoid
+
+variables {M N : Type*} [monoid M] [add_monoid N]
+
+/-- A submonoid of `M` is finitely generated if it is the closure of a finite subset of `M`. -/
+@[to_additive]
+def submonoid.fg (P : submonoid M) : Prop := ∃ S : finset M, submonoid.closure ↑S = P
+
+/-- An additive submonoid of `N` is finitely generated if it is the closure of a finite subset of
+`M`. -/
+add_decl_doc add_submonoid.fg
+
+variables (M N)
+
+/-- A monoid is finitely generated if it is finitely generated as a submonoid of itself. -/
+@[to_additive]
+def monoid.fg : Prop := (⊤ : submonoid M).fg
```

Comment:
What's the submodule version of this statement? Is it `module.finite`?

Reply:
Yes. This more general since it allows nonabelian monoid.

Reply:
Right; so should this be `class monoid.finite` instead of `def monoid.fg` to match?

Reply:
I made it into a `class`. I didn't change the name, since this is the standard terminology (we say finite module for a finitely generated module, but finite group means really finite, as set).

I also provided `add_monoid.fg` by hand since `to_additive` didn't work.